### PR TITLE
[Fix #44] Add meta testing; Update rebar.config to use rebar3

### LINKED ISSUE #44: Add meta testing; Update app.src for hex.pm; Move to Rebar3
esl/TODO#60

### PR DESCRIPTION
**Pull request by harenson**

### CODE REVIEW

File: src/elarm_alarmlist.erl [hunk 1]
```diff
@@ -151,36 +151,36 @@ al_test_() ->
...
 
 just_started() ->
-    {ok,State} = init([]),
-    ?assertEqual({{ok,[]}, State}, get_alarms(State)),
-    EvtId = erlang:now(),
-    ?assertEqual({{error, not_active},State}, get_alarm(EvtId, State)),
-    ?assertEqual({{error, not_active},State}, get_alarm(full, disk_1, State)).
+    {ok, State} = init([]),
+    ?assertEqual({{ok, []}, State}, get_alarms(State)),
+    EvtId = erlang:timestamp(),
```
Comment (hcs42) [inline src/elarm_alarmlist.erl]: We cannot use erlang:timestamp because that was introduced only in Erlang /OTP18, we need to support versions >= R15B03.

Comment (elbrujohalcon) [inline src/elarm_alarmlist.erl]: What about `os:timestamp()`, then?

Comment (hcs42) [inline src/elarm_alarmlist.erl]: The idea here is to generate a unique id. os:timestamp won't give a unique id.

File: src/elarm_server.erl [hunk 2]
```diff
@@ -367,7 +367,7 @@ create_alarm_rec(AlarmId, Src, AddInfo, Cfg) ->
 
 new_event_id() ->
     %% TODO: Fix this to something better
-    erlang:now().
+    erlang:timestamp().
```
Comment (hcs42) [inline src/elarm_server.erl]: Same here.

File: src/elarm_alarmlist.erl (hunk 1 above)
Comment (elbrujohalcon) [inline src/elarm_alarmlist.erl]: @hcs42 how do you plan to support OTP 18.x, then? Will you do something like…
```erlang
case erlang:function_exported(erlang, timestamp, 0) of
  true -> erlang:timestamp();
  false -> erlang:now()
end
```

Maybe you should consider switching to uuids.

Comment (hcs42) [inline src/elarm_alarmlist.erl]: As for now, OTP 18.2 still supports erlang:now (even thought it is deprecated), so this is not something I would address in this PR. 

As for the future, UUID would be more suitable than erlang:timestamp (which doesn't generate unique values as far as I understand). You are welcome to create a separate PR about that, e.g. by using http://github.com/travis/erlang-uuid.

Comment (elbrujohalcon) [inline src/elarm_alarmlist.erl]: Then, @harenson … you will have to somehow patch the meta suite (or maybe this module, by adding `-ignore_xref`and `-dialyzer` attributes) to remove the check for deprecated functions.

Comment (harenson) [inline src/elarm_alarmlist.erl]: @elbrujohalcon ! ok.

**Comment by elbrujohalcon** [comment]
@hcs42 this is ready for you to review it again, but I would bet travis errors are due to rebar3 not being installed on travis, right?

**Comment by hcs42** [comment]
Yes. Do you have a good way to fix that? (I guess so, since you are doing the same thing to all Inaka repos.)

**Comment by elbrujohalcon** [comment]
@hcs42 well… we'll need to download rebar3 first, then.
For our repos our plan is to consider rebar3 as an existing tool (since it will be provided with Erlang/OTP eventually) and don't worry about downloading it. As a matter of fact, rebar3 will be installed in [gadget](http://gadget.inakalabs.com) and [elvis](http://elvis.inakalabs.com) servers.

@harenson can you add that download part to travis.yml, please?

**Comment by harenson** [comment]
@elbrujohalcon Ok, I'll investigate about how to do that.

File: src/elarm_log.erl [hunk 3]
```diff
@@ -73,7 +73,8 @@ add_comment(_Alarm, _Comment, State) ->
     {ok, State}.
 
 %% Automatically clear an alarm
--spec clear(alarm(), clear_reason(), #log_state{}) -> {ok, #log_state{}} | {error, term()}.
+-spec clear(alarm(), clear_reason(), #log_state{}) -> {ok, #log_state{}} |
```
Comment (elvisci) [inline src/elarm_log.erl]: The spec in line 76 uses a record, please define a type for the record and use that instead.

File: src/elarm_log.erl [hunk 4]
```diff
@@ -50,11 +50,11 @@ init(_Opts) ->
...
 new_alarm(_Alarm, State) ->
-    {ok,State}.
+    {ok, State}.
 
 -spec repeat_alarm(alarm(), #log_state{}) -> {ok, #log_state{}}.
 repeat_alarm(_Alarm, State) ->
-    {ok,State}.
+    {ok, State}.
 
 -spec acknowledge(alarm(), ack_info(), #log_state{}) ->
```
Comment (elvisci) [inline src/elarm_log.erl]: The spec in line 59 uses a record, please define a type for the record and use that instead.

File: src/elarm_alarmlist.erl [hunk 5]
```diff
@@ -70,26 +72,26 @@ repeat_alarm(_Alarm, State) ->
...
               #al_state{ alarmlist = AList } = State) ->
     [{Key, Alarm}] = ets:lookup(AList, {AlarmId, Src}),
-    true = ets:insert(AList, {Key,Alarm#alarm{ ack_info = AckInfo,
-                                               state = new }}),
+    true = ets:insert(AList, {Key, Alarm#alarm{ ack_info = AckInfo,
+                                                state = new }}),
     {ok, State}.
 
 %% Add a comment to an alarm
 -spec add_comment(alarm_id(), alarm_src(), comment(), #al_state{}) ->
```
Comment (elvisci) [inline src/elarm_alarmlist.erl]: The spec in line 89 uses a record, please define a type for the record and use that instead.

File: src/elarm_event.erl [hunk 6]
```diff
@@ -72,7 +72,7 @@ unsubscribe(Ref, #evt_state{ subs = Subs } = State)->
 
 -spec new_alarm(alarm(), #evt_state{}) -> {ok, #evt_state{}} | {error, term()}.
```
Comment (elvisci) [inline src/elarm_event.erl]: The spec in line 73 uses a record, please define a type for the record and use that instead.

File: src/elarm_log.erl [hunk 7]
```diff
@@ -50,11 +50,11 @@ init(_Opts) ->
 
 -spec new_alarm(alarm(), #log_state{}) -> {ok, #log_state{}}.
 new_alarm(_Alarm, State) ->
-    {ok,State}.
+    {ok, State}.
 
 -spec repeat_alarm(alarm(), #log_state{}) -> {ok, #log_state{}}.
```
Comment (elvisci) [inline src/elarm_log.erl]: The spec in line 55 uses a record, please define a type for the record and use that instead.

File: src/elarm_event.erl [hunk 8]
```diff
@@ -50,19 +50,19 @@
...
+          {{ok, reference()}, #evt_state{}}.
 subscribe(Pid, Filter, #evt_state{ subs = Subs } = State)->
     MRef = monitor(process, Pid),
     Ref = make_ref(),
     NewSubs = add_subscriber(Ref, MRef, Pid, Filter, Subs),
     {{ok, Ref}, State#evt_state{ subs = NewSubs }}.
 
 %% Cancel subscription.
--spec unsubscribe(reference(), #evt_state{}) -> {ok,#evt_state{}}.
+-spec unsubscribe(reference(), #evt_state{}) -> {ok, #evt_state{}}.
```
Comment (elvisci) [inline src/elarm_event.erl]: The spec in line 62 uses a record, please define a type for the record and use that instead.

File: src/elarm_alarmlist.erl [hunk 9]
```diff
@@ -70,26 +72,26 @@ repeat_alarm(_Alarm, State) ->
 acknowledge(AlarmId, Src, AckInfo,
             #al_state{ alarmlist = AList } = State) ->
     [{Key, Alarm}] = ets:lookup(AList, {AlarmId, Src}),
-    true = ets:insert(AList, {Key,Alarm#alarm{ ack_info = AckInfo,
-                                               state = acknowledged }}),
+    true = ets:insert(AList, {Key, Alarm#alarm{ ack_info = AckInfo,
+                                                state = acknowledged }}),
     {ok, State}.
 
 -spec unacknowledge(alarm_id(), alarm_src(), ack_info(), #al_state{}) ->
```
Comment (elvisci) [inline src/elarm_alarmlist.erl]: The spec in line 79 uses a record, please define a type for the record and use that instead.

File: src/elarm_alarmlist.erl [hunk 10]
```diff
@@ -131,7 +133,7 @@ get_alarm(AlarmId, Src,
 -spec get_alarms(#al_state{}) ->
```
Comment (elvisci) [inline src/elarm_alarmlist.erl]: The spec in line 133 uses a record, please define a type for the record and use that instead.

File: src/elarm_config.erl [hunk 11]
```diff
@@ -81,25 +81,26 @@ get_unconfigured(#config_state{ missing_tab = MissingTab } = State) ->
 
 %% Get a list of all alarm_ids that have mapping configuration.
 -spec get_configured(#config_state{}) ->
```
Comment (elvisci) [inline src/elarm_config.erl]: The spec in line 83 uses a record, please define a type for the record and use that instead.

File: src/elarm_log.erl [hunk 12]
```diff
@@ -50,11 +50,11 @@ init(_Opts) ->
 
 -spec new_alarm(alarm(), #log_state{}) -> {ok, #log_state{}}.
```
Comment (elvisci) [inline src/elarm_log.erl]: The spec in line 51 uses a record, please define a type for the record and use that instead.

File: src/elarm_event.erl [hunk 13]
```diff
@@ -50,19 +50,19 @@
 init(_Opts) ->
     {ok, #evt_state{}}.
 
--spec subscribe(pid()|atom(), sub_filter(), #evt_state{}) ->
-          {{ok,reference()},#evt_state{}}.
+-spec subscribe(pid() | atom(), sub_filter(), #evt_state{}) ->
```
Comment (elvisci) [inline src/elarm_event.erl]: The spec in line 53 uses a record, please define a type for the record and use that instead.

File: src/elarm_config.erl [hunk 14]
```diff
@@ -81,25 +81,26 @@ get_unconfigured(#config_state{ missing_tab = MissingTab } = State) ->
...
+                    || {AlarmId} <- ets:tab2list(MissingTab)],
     Result = Configured ++ UnConfigured,
     {{ok, Result}, State}.
 
 get_default_configuration(#config_state{ default_map = Default } = State) ->
-    {{ok,Default}, State}.
+    {{ok, Default}, State}.
 
 %% Configure alarm mapping
 -spec add_configuration(alarm_id(), alarm_config(), #config_state{}) ->
```
Comment (elvisci) [inline src/elarm_config.erl]: The spec in line 106 uses a record, please define a type for the record and use that instead.

File: src/elarm_event.erl [hunk 15]
```diff
@@ -72,7 +72,7 @@ unsubscribe(Ref, #evt_state{ subs = Subs } = State)->
 
 -spec new_alarm(alarm(), #evt_state{}) -> {ok, #evt_state{}} | {error, term()}.
 new_alarm(Alarm, #evt_state{ subs = Subs } = State) ->
-    send_events(Alarm, Subs),
+    _ = send_events(Alarm, Subs),
     {ok, State}.
 
 -spec repeat_alarm(alarm(), #evt_state{}) ->
```
Comment (elvisci) [inline src/elarm_event.erl]: The spec in line 78 uses a record, please define a type for the record and use that instead.

**Comment by benoitc** [comment]
bump?

File: .travis.yml [hunk 16]
```diff
@@ -9,4 +9,7 @@ otp_release:
    - R15B03
    - R15B02
    - R15B01
-script: "make all test doc"
+
+script:
```
Comment (bryanhuntesl) [inline .travis.yml]: No longer necessary - Travis provide rebar3 by default now

**Comment by bryanhuntesl** [comment]
This is going nowhere - creating a new pull request - things are simpler now to get working - rebar3 comes installed on travis.org - actually it's quite trivial.

**Pull request closed without merging**